File: app/src/main/java/com/example/contactsviewer/ui/ContactViewModel.java
package com.example.contactsviewer.ui;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.contactsviewer.data.DummyRepository;
import com.example.contactsviewer.model.Contact;

import java.util.List;

public class ContactViewModel extends AndroidViewModel {
    MutableLiveData<List<Contact>> contacts = new MutableLiveData<>();
    private final DummyRepository repository;

    public ContactViewModel(Application application) {
        super(application);
        repository = new DummyRepository(application);
        loadContacts();
    }

    private void loadContacts() {
        contacts.setValue(repository.getContacts());
    }

    public LiveData<List<Contact>> getContacts() {
        return contacts;
    }
}

File: app/src/main/java/com/example/contactsviewer/ui/ContactViewModel.kt
package com.example.contactsviewer.ui

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.contactsviewer.data.ContactRepository
import com.example.contactsviewer.model.Contact
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class ContactViewModel(private val repository: ContactRepository) : ViewModel() {

    val contacts: StateFlow<List<Contact>> = repository.getAllContacts()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    fun addContact(context: Context, contact: Contact, selectedImageUri: Uri?) {
        viewModelScope.launch {
            val newContactId = repository.addContact(contact)
            selectedImageUri?.let { uri ->
                saveAvatarImage(context, uri, newContactId)
            }
        }
    }

    fun deleteContact(contactId: Long) {
        viewModelScope.launch {
            repository.deleteContact(contactId)
        }
    }

    fun saveAvatarImage(context: Context, uri: Uri, contactId: Long) {
        viewModelScope.launch {
            repository.saveAvatarImage(context, uri, contactId)
        }
    }
}
